### PR TITLE
JS3 Sprint 4 prep: Promises, then, async/await, try/catch, getFilms

### PR DESCRIPTION
**Pull request by SallyMcGrath**
## What does this change?

Module: JS3
Week(s): 4

## Checklist

- [x] I have read the [contributing guidelines](CONTRIBUTING.MD)
- [x] I have checked my spelling and grammar with an [automated tool](https://www.grammarly.com/grammar-check)
- [x] I have previewed my changes to check the [markdown renders](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) as I intend
- [x] I have run my code to check it works
- [x] My changes follow our [Style Guide](https://curriculum.codeyourfuture.io/guides/code-style-guide)

## Description

Completing the concept map from JS3-3 with the final parts on Promises, then, async/await, and try/catch. I brought filterFilms back from week 2 and updated it with a fetch as the last part.

Please do checkout this PR and edit directly as well as offering line edits. If we can simplify or reduce, let's do this!

I will go away and fix the horrible Mermaid in solo view bug after React is ported. For now just reload the page (sorry!!!)

### CODE REVIEW

File: content/en/js3/blocks/promises/index.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+```mermaid
+graph LR
+    Asynchrony --> |delivered with| Promise{{🤝 Promises}}
+    Promise --> |resolve to a| Response{📤 Response}
+    Promise --> |join the| EventLoop{{Event Loop 🔁}}
+```
+
+To get data from a server, we make a request with `fetch`. We act on what comes back: the response. But what happens in the middle? We already know that JavaScript is single-threaded: it can only do one thing at a time.
+
+So do we just stop and wait? No! We have a special object to handle this time problem. Run this code in your terminal:
```
Comment (illicitonion) [inline content/en/js3/blocks/promises/index.md:21]: Is this asking them to run this in `node` or in the web inspector console?

File: content/en/js3/blocks/then/index.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    .then((htmlUrl) => console.log(htmlUrl));
+};
+getProfile("https://api.github.com/users/SallyMcGrath");
+```
+
+So `then` returns a _new_ `Promise`, and you can call `then` again on the new object. You can chain Promises in ever more complex dependent steps. This is called [Promise chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining).
+
+It's important to understand some of what is happening with Promises and `then`. But for the most part, you will not be writing code in this style.
+
+You will be writing ([declaring](/guides/paradigm/)) a description of what you want to happen and a [loosely-coupled](https://en.wikipedia.org/wiki/Loose_coupling) set of stateless, single-purpose functions to make it so. To help you stay thinking in this way, let's explore async/await.
```
Comment (illicitonion) [inline content/en/js3/blocks/then/index.md:56]: This line feels a bit abstract and not very explained... I think we should either expand it or remove it - which do you prefer?

File: content/en/js3/blocks/async-await/index.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
++++
+
+```mermaid
+graph LR
+    Promise{{🤝 Promises}} --> |syntax| async{{🏃‍♂️ async}}
+    async --> |syntax| await{{📭 await}}
+    await --> |resolves to| Response{{📤 Response}}
+```
+
+Async/await is {{<tooltip title="syntactic sugar">}}A simpler, or "sweeter" way to write the same thing. The code works the same under the hood, but it's easier to read. {{</tooltip>}} for Promises. We group them together: async/await, because we {{<tooltip title="use them together. ">}}We can only use `await` inside an `async` function or at the [top level](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await) of a module.{{</tooltip>}}
```
Comment (illicitonion) [inline content/en/js3/blocks/async-await/index.md:19]: AFAICT it's not possible to focus on the tooltip, so the "top level" link here isn't clickable - when I move the mouse to the tooltip, it gets hidden.

File: content/en/js3/blocks/async-await/index.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+We use the [`async`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) [keyword](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#keywords) to define a function that returns a Promise. An async function always returns a Promise.
```
Comment (illicitonion) [inline content/en/js3/blocks/async-await/index.md:21]: For some reason the keywords anchor doesn't actually scroll to the right place in this page for me 🤷

File: content/en/js3/blocks/catch/index.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+headless = true
+time = 10
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Define a try/catch block'
+    2='Handle an error'
++++
+
+We can handle errors with a try/catch block. We can use the `try` keyword to _try_ to do something, and if it fails, `catch` the {{<tooltip title="error">}}An [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) is a global object produced when something goes wrong. We can throw an [Error](https://www.w3schools.com/js/js_errors.asp) manually with the `throw` keyword.{{</tooltip>}} We can use try/catch in both synchronous and asynchronous code.
```
Comment (illicitonion) [inline content/en/js3/blocks/catch/index.md:12]: Again, tooltip links don't work?

File: content/en/js3/blocks/fetch-films/index.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+{{</tab>}}
+{{</tabs>}}
+
+We've added a `try...catch` block to handle any errors that might occur. We've also added `await` to the `fetch` and `response.json()` calls. This means that the function will sensibly wait for the `fetch` to complete before trying to get the json data from the response.
+
+In our last implementation, we called the render function straight away. This time, we need to wait for the films to be fetched before we can render them. Write a new async function to initialise our app. Try to write it yourself first, then check your understanding below.
+
+<details>
+<summary>Your `init` function should look something like this:</summary>
```
Comment (illicitonion) [inline content/en/js3/blocks/fetch-films/index.md:63]: Markdown doesn't render in `<summary>` tags unless they have an empty newline at the start of them, which means the `init` is a literal \`init\`... See DM for more details

**Review by illicitonion — COMMENTED** [review]
Generally looks good, thanks for putting this together! I opened https://github.com/CodeYourFuture/curriculum/pull/486 with a few edits.

A few of other high-level thoughts to discuss:
1. It feels like we should've introduced `try`/`catch` ´before now. Any thoughts on where we should insert it?
2. I'd quite like us to lean into error handling a bit more - currently most of our trainees think that `console.error` is suitable error handling, but I'd like us to start guiding people towards thinking "How can I show the user that something went wrong, and what they can/should do about it?" - any thoughts as to where/how/when to start doing this?
3. Do we want to talk about errors in _Promises_ at all (i.e. reject and/or talk a bit more about what the `console.log`'d `Promise` values we show are), or are we happy to just say "use `async`/`await` and `try`/`catch`?

File: content/en/js3/blocks/promises/index.md (hunk 1 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/promises/index.md:21]: Yes you can almost see me realising they haven't done try/catch in the text, can't you? 😂 

I had them running stuff in the browser but then I realised that has a variance we cannot control: CORS settings on the website apply to the console so if they open their console on MDN they will get different results to W3Schools or us. So then I moved it all to the REPL.  Let's specify REPL throughout

File: content/en/js3/blocks/then/index.md (hunk 2 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/then/index.md:56]: Yeah, I agree.

Here's my thinking here: at about this point I often get into long discussions with trainees about why we want them to use map, why we keep harassing them to break up their functions, etc. They don't get the point of it and they feel annoyed by it, like we're deliberately making things harder just to vex them. My best response to this has generally been to explain we're learning a style of programming, that they can go read about if they like. And that there are other styles and that's fine, but we're doing this for now. JS3 is IME when these points reach a crescendo, so I'd like to address it. What do you think? How could you rephrase?

File: content/en/js3/blocks/async-await/index.md (hunk 3 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/async-await/index.md:19]: Oh hm, I can fix that -- I habitually tab so I didn't notice. Make a ticket?

File: content/en/js3/blocks/async-await/index.md (hunk 4 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/async-await/index.md:21]: I KNOW IT IS SO ANNOYING

File: content/en/js3/blocks/async-await/index.md (hunk 3 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/async-await/index.md:19]: #490 there you go

File: content/en/js3/blocks/fetch-films/index.md (hunk 6 above)
Comment (SallyMcGrath) [reply content/en/js3/blocks/fetch-films/index.md:63]: #488  / #489  💥

**Comment by SallyMcGrath** [comment]
> Generally looks good, thanks for putting this together! I opened #486 with a few edits.
> 
> A few of other high-level thoughts to discuss:
> 
> 1. It feels like we should've introduced `try`/`catch` ´before now. Any thoughts on where we should insert it?

JS2 - 4 seems the obvious place: https://curriculum.codeyourfuture.io/js2/sprints/4/prep/   IT EMPTY 🕳️ 

> 2. I'd quite like us to lean into error handling a bit more - currently most of our trainees think that `console.error` is suitable error handling, but I'd like us to start guiding people towards thinking "How can I show the user that something went wrong, and what they can/should do about it?" - any thoughts as to where/how/when to start doing this?

So I would say a small piece in JS2 - 4 and then you can have a max 45 minute workshop in JS3 - 4. I had planned this workshop to be web components but I think it's too much and they need the time to do their project. 

> 3. Do we want to talk about errors in _Promises_ at all (i.e. reject and/or talk a bit more about what the `console.log`'d `Promise` values we show are), or are we happy to just say "use `async`/`await` and `try`/`catch`?

I think we should move swiftly on from Promises and let them explore these things when they're a bit more confident reading the code and mentally modelling it. 

My main quibble with what I've written here is I've consistently collapsed the event loop and the task queues. I say join the event loop, or queue to the event loop, when it's more accurate to say join the queue. I did this because there are different queues and I didn't want to get into it, but now I wonder if saying join the event loop isn't even more obscuring. What do you think?

File: content/en/js3/blocks/then/index.md (hunk 2 above)
Comment (illicitonion) [reply content/en/js3/blocks/then/index.md:62]: I'm inclined to say: Let's remove this from the prep, and try to insert a 30 minute workshop into the day to actively have the discussion (which we can maybe try to put together in the future, through some discussion, rather than trying to make Right Now)?

I think this is something that it's hard to really understand the value of without working through examples (and I'm slightly skeptical that with the current course content we can even motivate that very well - really it's more long-term projects, approaching others' code, and testability where you start seeing the values of this)...

Let's maybe discuss synchronously how these discussions tend to go?

**Comment by illicitonion** [comment]
> > Generally looks good, thanks for putting this together! I opened #486 with a few edits.
> > A few of other high-level thoughts to discuss:
> > 
> > 1. It feels like we should've introduced `try`/`catch` ´before now. Any thoughts on where we should insert it?
> 
> JS2 - 4 seems the obvious place: https://curriculum.codeyourfuture.io/js2/sprints/4/prep/ IT EMPTY 🕳️

Filed https://github.com/CodeYourFuture/curriculum/issues/495

> > 2. I'd quite like us to lean into error handling a bit more - currently most of our trainees think that `console.error` is suitable error handling, but I'd like us to start guiding people towards thinking "How can I show the user that something went wrong, and what they can/should do about it?" - any thoughts as to where/how/when to start doing this?
> 
> So I would say a small piece in JS2 - 4 and then you can have a max 45 minute workshop in JS3 - 4. I had planned this workshop to be web components but I think it's too much and they need the time to do their project.

Sounds good - let's flesh out that discussion on the issue above.

> > 3. Do we want to talk about errors in _Promises_ at all (i.e. reject and/or talk a bit more about what the `console.log`'d `Promise` values we show are), or are we happy to just say "use `async`/`await` and `try`/`catch`?
> 
> I think we should move swiftly on from Promises and let them explore these things when they're a bit more confident reading the code and mentally modelling it.

Sounds good, let's leave this in exactly the level of detail you've gone into - we can see how this goes with the first few classes and work out if there's anything we want to change from there.

> My main quibble with what I've written here is I've consistently collapsed the event loop and the task queues. I say join the event loop, or queue to the event loop, when it's more accurate to say join the queue. I did this because there are different queues and I didn't want to get into it, but now I wonder if saying join the event loop isn't even more obscuring. What do you think?

I think this is a completely fine level of detail for where our trainees are and what they're trying to do. These distinctions are _interesting_ but feel quite internal implementation detailsy of JS engines rather than something someone needs to fully understand in order to be productive or write high-quality code?

**Review by illicitonion — APPROVED** [review]
I pushed a commit with the last few edits, including removing the declarative sentence - feel free to either merge or push back :)
